### PR TITLE
first attempt at using SnoopCompile to speed things up

### PR DESCRIPTION
**Pull request by palday**
## current main

```bash
~/MixedModels.jl$ for i in {1..5}; do echo "Run $i"; julia --project=. -e'using MixedModels; @time fit(MixedModel, @formula(reaction ~ 1 + days + (1+days|subj)), MixedModels.dataset(:sleepstudy)); @time fit(MixedModel,@formula(use ~ 1+age+abs2(age)+urban+livch+(1|urban&dist)), MixedModels.dataset(:contra), Bernoulli(), LogitLink())'; done
Run 1
[ Info: using MKL.jl
Minimizing 58    Time: 0:00:00 (13.88 ms/it)
 22.300008 seconds (77.44 M allocations: 4.614 GiB, 6.05% gc time, 99.83% compilation time)
Minimizing 254   Time: 0:00:00 ( 1.10 ms/it)
 16.647709 seconds (49.34 M allocations: 2.975 GiB, 4.31% gc time, 99.16% compilation time)
Run 2
[ Info: using MKL.jl
Minimizing 58    Time: 0:00:00 (12.15 ms/it)
 22.237959 seconds (77.44 M allocations: 4.614 GiB, 5.52% gc time, 99.84% compilation time)
Minimizing 254   Time: 0:00:00 ( 1.27 ms/it)
 16.607973 seconds (49.34 M allocations: 2.975 GiB, 4.44% gc time, 98.91% compilation time)
Run 3
[ Info: using MKL.jl
Minimizing 58    Time: 0:00:00 (12.33 ms/it)
 21.382303 seconds (77.44 M allocations: 4.614 GiB, 5.54% gc time, 99.84% compilation time)
Minimizing 254   Time: 0:00:00 ( 1.09 ms/it)
 16.072365 seconds (49.34 M allocations: 2.975 GiB, 4.35% gc time, 99.13% compilation time)
Run 4
[ Info: using MKL.jl
Minimizing 58    Time: 0:00:00 (12.55 ms/it)
 21.785942 seconds (77.44 M allocations: 4.614 GiB, 5.47% gc time, 99.83% compilation time)
Minimizing 254   Time: 0:00:00 ( 1.12 ms/it)
 16.510533 seconds (49.34 M allocations: 2.975 GiB, 4.26% gc time, 99.15% compilation time)
Run 5
[ Info: using MKL.jl
Minimizing 58    Time: 0:00:00 (12.30 ms/it)
 22.517317 seconds (77.44 M allocations: 4.614 GiB, 5.32% gc time, 99.84% compilation time)
Minimizing 254   Time: 0:00:00 ( 1.19 ms/it)
 16.357725 seconds (49.34 M allocations: 2.975 GiB, 4.41% gc time, 99.14% compilation time)
```

## with new precompile
```bash
~/MixedModels.jl$ for i in {1..5}; do echo "Run $i"; julia --project=. -e'using MixedModels; @time fit(MixedModel, @formula(reaction ~ 1 + days + (1+days|subj)), MixedModels.dataset(:sleepstudy)); @time fit(MixedModel,@formula(use ~ 1+age+abs2(age)+urban+livch+(1|urban&dist)), MixedModels.dataset(:contra), Bernoulli(), LogitLink())'; done
[ Info: using MKL.jl
Minimizing 58    Time: 0:00:00 ( 6.42 ms/it)
 13.022910 seconds (42.46 M allocations: 2.437 GiB, 4.60% gc time, 99.66% compilation time)
Minimizing 254   Time: 0:00:00 ( 1.15 ms/it)
 12.423930 seconds (38.11 M allocations: 2.248 GiB, 5.15% gc time, 98.81% compilation time)
Run 2
[ Info: using MKL.jl
Minimizing 58    Time: 0:00:00 ( 6.04 ms/it)
 12.812662 seconds (42.46 M allocations: 2.437 GiB, 4.50% gc time, 99.72% compilation time)
Minimizing 254   Time: 0:00:00 ( 1.20 ms/it)
 12.360282 seconds (38.11 M allocations: 2.248 GiB, 5.20% gc time, 98.79% compilation time)
Run 3
[ Info: using MKL.jl
Minimizing 58    Time: 0:00:00 ( 6.42 ms/it)
 14.338150 seconds (42.46 M allocations: 2.437 GiB, 4.70% gc time, 99.75% compilation time)
Minimizing 254   Time: 0:00:00 ( 1.19 ms/it)
 13.086941 seconds (38.11 M allocations: 2.248 GiB, 5.23% gc time, 98.88% compilation time)
Run 4
[ Info: using MKL.jl
Minimizing 58    Time: 0:00:00 ( 6.24 ms/it)
 13.455821 seconds (42.46 M allocations: 2.437 GiB, 4.52% gc time, 99.72% compilation time)
Minimizing 254   Time: 0:00:00 ( 1.53 ms/it)
 14.986385 seconds (38.11 M allocations: 2.248 GiB, 5.07% gc time, 98.75% compilation time)
Run 5
[ Info: using MKL.jl
Minimizing 58    Time: 0:00:00 ( 6.35 ms/it)
 12.882482 seconds (42.46 M allocations: 2.437 GiB, 4.56% gc time, 99.71% compilation time)
Minimizing 254   Time: 0:00:00 ( 1.16 ms/it)
 12.763560 seconds (38.12 M allocations: 2.248 GiB, 5.20% gc time, 98.88% compilation time)
```

## how I did this
```julia
using SnoopCompile, MixedModels
ttot, pcs = SnoopCompile.parcel(snoop)
snoop = @snoopi_deep  begin 
       MixedModels.fit(MixedModel, @formula(reaction ~ 1 + days + (1+days|subj)), MixedModels.dataset(:sleepstudy))
       MixedModels.fit(MixedModel,@formula(use ~ 1+age+abs2(age)+urban+livch+(1|urban&dist)), MixedModels.dataset(:contra), Bernoulli())
       end
SnoopCompile.write(joinpath(@__DIR__, "src"), pcs;has_bodyfunction=true)
```

I then had to edit `precompile_MixedModels.jl` and replace a few type parameters that were still there (`s/T<:Core.AbstractFloat/Float64/g` and `s/D<:Distributions.Distribution/Distributions.Bernoulli/g`).

Because of the way Tables.jl and StatsModels use NamedTuples, I also had to remove lines referencing specific variables in these models.

## PR template
Did behavior change? Did you add need features? If so, please update NEWS.md
- [x] add entry in NEWS.md
- [x] after opening this PR, add a reference and run `docs/NEWS-update.jl` to update the cross-references.

Should we release your changes right away? If so, bump the version:
- [x] I've bumped the version appropriately

### CODE REVIEW

**Comment by palday** [comment]
@dmbates should we turn off autoformatting for the precompile files? They're generated by SnoopCompile and if we leave them as emitted then it makes it easier to compare things when we do more inspection with Snoop.

File: src/precompile_MixedModels.jl [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    Base.precompile(Tuple{typeof(deviance!),GeneralizedLinearMixedModel,Int64})   # time: 0.6433022
+    Base.precompile(Tuple{Type{LinearMixedModel},Vector{Float64},FeTerm{Float64, Matrix{Float64}},Vector{AbstractReMat{Float64}},FormulaTerm{ContinuousTerm{Float64}, Tuple{MatrixTerm{Tuple{InterceptTerm{true}, ContinuousTerm{Float64}}}, RandomEffectsTerm}},Vector{Any},Nothing})   # time: 0.42538697
+    Base.precompile(Tuple{typeof(deviance!),GeneralizedLinearMixedModel{Float64, Bernoulli{Float64}},Int64})   # time: 0.3716886
+    Base.precompile(Tuple{typeof(_ranef_refs),InteractionTerm{Tuple{CategoricalTerm{StatsModels.FullDummyCoding, String, 2}, CategoricalTerm{StatsModels.FullDummyCoding, String, 60}}},NamedTuple{(:use, :age, :urban, :livch, :dist), Tuple{Vector{String}, Vector{Float64}, Vector{String}, Vector{String}, Vector{String}}}})   # time: 0.15686166
+    Base.precompile(Tuple{typeof(fit!),GeneralizedLinearMixedModel{Float64, Bernoulli{Float64}}})   # time: 0.1559177
+    Base.precompile(Tuple{typeof(modelcols),RandomEffectsTerm,NamedTuple{(:reaction, :days, :subj), Tuple{Vector{Float64}, Vector{Int8}, Vector{String}}}})   # time: 0.14000398
```
Comment (ararslan) [inline src/precompile_MixedModels.jl:25]: Looks like this line and 44 are overly specific to the calls used to generate this, as the particular `NamedTuple` table structure is part of the method signature that's being precompiled. When testing the timing improvements I would try with a different dataset to see how much of an impact there is.

I've never used SnoopCompile before; are the timings provided in the comments at the end of each line the amount of time it took to execute the `precompile` call? If so, we're adding 0.157 seconds to precompilation time that will only benefit two specific examples. Obviously not a ton of time but figured it's worth noting.

Comment (palday) [reply src/precompile_MixedModels.jl:25]: Good catch! And super interesting to see how the Tables.jl representation plays through these things. I had also noticed that SnoopCompile had a lot to say about StatsModels that was quite specific, but then again StatsModels is known for hitting the compiler hard.

**Comment by dmbates** [comment]
> @dmbates should we turn off autoformatting for the precompile files? They're generated by SnoopCompile and if we leave them as emitted then it makes it easier to compare things when we do more inspection with Snoop.

Agreed.

**Review by dmbates — APPROVED** [review]
Should we update the compat entry for `StatsFuns` in `Project.toml` as suggested in #607 while doing this?

**Review by ararslan — APPROVED** [review]

**Comment by palday** [comment]
Here some final timings using `benchmarks/snoopspeed.sh` (which uses the kb07 and verbagg datasets, so is more of a test of generalization than the previous tests):

## current main
```bash
~/MixedModels.jl$ time benchmark/snoopspeed.sh
Run 1
[ Info: using MKL.jl
Minimizing 93    Time: 0:00:01 (13.39 ms/it)
 25.721185 seconds (88.90 M allocations: 5.222 GiB, 5.69% gc time, 99.73% compilation time)
[ Info: using MKL.jl
Minimizing 203   Time: 0:00:01 ( 5.06 ms/it)
 34.859563 seconds (117.98 M allocations: 7.119 GiB, 5.12% gc time, 97.94% compilation time)
Run 2
[ Info: using MKL.jl
Minimizing 93    Time: 0:00:01 (12.92 ms/it)
 25.323026 seconds (88.90 M allocations: 5.222 GiB, 5.24% gc time, 99.77% compilation time)
[ Info: using MKL.jl
Minimizing 203   Time: 0:00:01 ( 5.09 ms/it)
 35.383858 seconds (117.98 M allocations: 7.119 GiB, 5.17% gc time, 97.96% compilation time)
Run 3
[ Info: using MKL.jl
Minimizing 93    Time: 0:00:01 (16.60 ms/it)
 26.033828 seconds (88.90 M allocations: 5.222 GiB, 5.14% gc time, 99.77% compilation time)
[ Info: using MKL.jl
Minimizing 203   Time: 0:00:01 ( 5.02 ms/it)
 37.257851 seconds (117.98 M allocations: 7.119 GiB, 5.12% gc time, 98.06% compilation time)

real    3m40.261s
user    4m2.498s
sys     0m7.523s
```

## this PR
```bash
~/MixedModels.jl$ time benchmark/snoopspeed.sh
Run 1
[ Info: using MKL.jl
Minimizing 93    Time: 0:00:01 (12.58 ms/it)
 19.097108 seconds (55.30 M allocations: 3.117 GiB, 3.88% gc time, 99.67% compilation time)
[ Info: using MKL.jl
Minimizing 203   Time: 0:00:01 ( 5.41 ms/it)
 25.843597 seconds (75.05 M allocations: 4.380 GiB, 4.62% gc time, 97.07% compilation time)
Run 2
[ Info: using MKL.jl
Minimizing 93    Time: 0:00:00 ( 9.59 ms/it)
 17.110819 seconds (55.30 M allocations: 3.117 GiB, 4.38% gc time, 99.65% compilation time)
[ Info: using MKL.jl
Minimizing 203   Time: 0:00:01 ( 5.52 ms/it)
 23.983724 seconds (75.05 M allocations: 4.380 GiB, 4.67% gc time, 96.65% compilation time)
Run 3
[ Info: using MKL.jl
Minimizing 93    Time: 0:00:00 ( 9.28 ms/it)
 16.480201 seconds (55.30 M allocations: 3.117 GiB, 4.36% gc time, 99.66% compilation time)
[ Info: using MKL.jl
Minimizing 203   Time: 0:00:01 ( 5.29 ms/it)
 24.536892 seconds (75.05 M allocations: 4.380 GiB, 4.80% gc time, 96.97% compilation time)

real    3m4.099s
user    3m27.271s
sys     0m6.946s
```

**Pull request merged**